Fix `fetch` response stream cancellation

simolus3 opened the pull request:
The `fetch` client turns a `ReadableStreamDefaultReader` into a Dart `Stream` by repeatedly invoking `reader.read()` in an `async*` function.

This generally works well, but a problem with `async*` functions is that they can only respect cancellations when yielding. This results in cancellations taking longer than they need to in the following case:

1. A chunk of data is reported from the reader.
2. This is emitted by the stream, and we go into the next iteration of the `_readBody` loop.
3. Asynchronously, the listener calls `.cancel()` on its subscription.
4. This now has to wait for the next chunk of the source stream instead of aborting the request.

Since we already have an `AbortController` set up, this fixes the issue by adding an `onCancel` callback to abort the `fetch` request when the stream subscription is cancelled.

Closes https://github.com/dart-lang/http/issues/1820.

cc @brianquinlan 

---

- [X] I’ve reviewed the contributor guide and applied the relevant portions to this PR.

<details>
  <summary>Contribution guidelines:</summary><br>

- See our [contributor guide](https://github.com/dart-lang/.github/blob/main/CONTRIBUTING.md) for general expectations for PRs.
- Larger or significant changes should be discussed in an issue before creating a PR.
- Contributions to our repos should follow the [Dart style guide](https://dart.dev/guides/language/effective-dart) and use `dart format`.
- Most changes should add an entry to the changelog and may need to [rev the pubspec package version](https://github.com/dart-lang/sdk/blob/main/docs/External-Package-Maintenance.md#making-a-change).
- Changes to packages require [corresponding tests](https://github.com/dart-lang/.github/blob/main/CONTRIBUTING.md#Testing).

Note that many Dart repos have a weekly cadence for reviewing PRs - please allow for some latency before initial review feedback.
</details>

brianquinlan commented on pkgs/http/lib/src/browser_client.dart:
```diff
@@ -155,10 +162,27 @@ Never _rethrowAsClientException(Object e, StackTrace st, BaseRequest request) {
-Stream<List<int>> _readBody(BaseRequest request, Response response) async* {
+Stream<List<int>> _readBody(BaseRequest request, Response response,
+    bool Function() subscriptionWasCancelled) async* {
+  void rethrowAsClientException(Object e, StackTrace st) {
```
This could just be a matter of taste, but I find the approach of using request abort to handle stream cancellation problematic. Could I suggest that you try a different approach so that we can compare? That approach would be to create a `StreamController` in this method and then do chained calls to `read`. Pseudocode:

```dart
var Completer? readDoneOrCancelled;
final s = StreamController(
  ...
  onCancel: () { isDone = true; readDoneOrCancelled?.complete() };
);
while (!isDone) {
    readDoneOrCancelled = Completer();
    final chunk = bodyStreamReader.read().toDart.then((chunk) {
      readDoneOrCancelled.complete();
      if (chunk.done) {
        isDone = true;
      }
      s.add(chunk.value! as JSUint8Array).toDart);
    });
  await readDoneOrCancelled.future;
}

```

simolus3 replied:
I like that approach :+1: We also need to handle the stream pausing but that doesn't make this much more complicated either.

brianquinlan reviewed (commented):
Very high quality PR! Especially the added conformance test (a few of which seem to be failing).

simolus3 commented:
It looks like the remaining failures are all infra or appear to be unrelated to my changes. Let me know if there's anything I should look into.

lrhn commented on pkgs/http/lib/src/browser_client.dart:
```diff
@@ -155,50 +155,105 @@ Never _rethrowAsClientException(Object e, StackTrace st, BaseRequest request) {
+
+      controller.add((chunk.value! as JSUint8Array).toDart);
+      if (controller.isPaused) {
+        final resume = waitingForResume = Completer();
```
(Good choice to only allocate completer if actually needed, and not for each pause/resume!)

simolus3 commented on pkgs/http/lib/src/browser_client.dart:
```diff
@@ -155,50 +155,105 @@ Never _rethrowAsClientException(Object e, StackTrace st, BaseRequest request) {
+      await pipeIntoController;
+    };
+
+  return controller.stream;
```
Both MDN and (from a quick look) also the streams whatwg standard are super unclear on whether we need to call `.cancel()` on the underlying `ReadableStreamDefaultReader` if the reader emits a done event. Intuitively I'd say it's not necessary, but the old implementation did it and I'm not sure if this may be a leak if we don't do it.

simolus3 replied:
Hm no, judging by step 2 of [this spec](https://streams.spec.whatwg.org/#readable-stream-cancel), it appears to be unecessary. That information is hard to find though, I'll add a comment somewhere.

simolus3 commented:
Alright, I have adopted most of the suggestions shared by @lrhn. I think they make the control flow a lot more obvious 👍 

I think the correct behavior (which we also had with the old `async*` implementation) is to only call `getReader()` when the stream is actually listened to, so I've moved most of the setup into the `onListen` callback. That also avoids the use of finalizers here.

Despite being a single-use stream, I think using `Stream.multi` may be helpful here to be explicit about adding events synchronously? I've also added a comment on each `addSync` / `addErrorSync` / `closeSync` invocation to explain why it's safe.

brianquinlan commented on pkgs/http/lib/src/browser_client.dart:
```diff
@@ -155,49 +156,98 @@ Never _rethrowAsClientException(Object e, StackTrace st, BaseRequest request) {
+          // If the stream enters an error state, .cancel() will rethrow that
+          // error. We want to ignore that since we would have added it to the
+          // stream already.
+          _rethrowAsClientException(e, s, request);
```
This will appear as an asynchronous exception, right? I really hate those.

Also, is it possible for it to throw. The [mdn](https://developer.mozilla.org/en-US/docs/Web/API/ReadableStreamDefaultReader/cancel) says:

> Exceptions
> TypeError
> The source object is not a ReadableStreamDefaultReader, or the stream has no owner.

If you change this, you should also change the comment on line 217.

brianquinlan commented on pkgs/http_client_conformance_tests/lib/src/response_body_streamed_test.dart:
```diff
@@ -77,5 +77,63 @@ void testResponseBodyStreamed(Client client,
       });
       await cancelled.future;
     });
+
```
[optional] Would you mind adding some tests to check that stream pause and resume works?

simolus3 replied:
I have added a few tests ensuring pause and resume work. Let me know if you were looking for more. To properly test this, one would probably have to check that pausing a stream on the client eventually also pauses the stream on the server due to flow control. But that seems like a very tricky test to write.

brianquinlan commented:
It looks like one test is failing on WASM.

simolus3 commented:
Thanks, I've fixed that test.

brianquinlan commented:
The remaining CI failures are unrelated.

simolus3 commented:
Thanks for fixing the test and reviewing this!
